velox/expression/VectorReaders.h

```diff
@@ -98,40 +99,78 @@ struct VectorReader {
   const DecodedVector& decoded_;
 };
 
-// This VectorReader is optimized for primitive types in constant or flat
-// encoded vectors.  It operates directly on the vector's content avoiding
-// the need to go through the expensive decoding process.
+// ConstantVectorReader and FlatVectorReader are optimized for primitive types
+// in constant or flat encoded vectors.  They operate directly on the vector's
+// content avoiding the need to go through the expensive decoding process.
 template <typename T>
-struct ConstantFlatVectorReader {
+struct ConstantVectorReader {
   using exec_in_t = typename VectorExec::template resolver<T>::in_type;
 
-  explicit ConstantFlatVectorReader(const FlatVector<exec_in_t>* vector)
-      : rawValues_(vector->rawValues()),
-        rawNulls_(vector->rawNulls()),
-        indexMultiple_(1) {}
+  std::optional<exec_in_t> value;
+  explicit ConstantVectorReader<T>(
+      const std::shared_ptr<facebook::velox::BaseVector>& vector) {
+    if (!vector->isNullAt(0)) {
+      value = *vector->asUnchecked<ConstantVector<exec_in_t>>()->rawValues();
+    }
+  }
 
-  explicit ConstantFlatVectorReader(const ConstantVector<exec_in_t>* vector)
-      : rawValues_(vector->rawValues()),
-        rawNulls_(vector->isNullAt(0) ? &bits::kNull64 : nullptr),
-        indexMultiple_(0) {}
+  exec_in_t operator[](vector_size_t) const {
+    return *value;
+  }
 
-  explicit ConstantFlatVectorReader(const VectorReader<T>&) = delete;
-  VectorReader<T>& operator=(const VectorReader<T>&) = delete;
+  exec_in_t readNullFree(vector_size_t) const {
+    return *value;
+  }
+
+  bool isSet(vector_size_t) const {
+    return value.has_value();
+  }
+
+  bool mayHaveNulls() const {
+    return !value.has_value();
+  }
+
+  inline bool containsNull(vector_size_t) const {
+    return !value.has_value();
+  }
+
+  bool containsNull(vector_size_t, vector_size_t) const {
+    return !value.has_value();
+  }
+
+  inline bool mayHaveNullsRecursive() const {
+    return !value.has_value();
+  }
+
+  // Scalars don't have children, so this is a no-op.
+  void setChildrenMayHaveNulls() {}
+};
+
+template <typename T>
+struct FlatVectorReader {
+  using exec_in_t = typename VectorExec::template resolver<T>::in_type;
+  const exec_in_t* values;
+  FlatVector<exec_in_t>* vector;
+
+  explicit FlatVectorReader<T>(
+      const std::shared_ptr<facebook::velox::BaseVector>& baseVector)
+      : values(baseVector->asUnchecked<FlatVector<exec_in_t>>()->rawValues()),
+        vector(baseVector->asUnchecked<FlatVector<exec_in_t>>()) {}
 
   exec_in_t operator[](vector_size_t offset) const {
-    return rawValues_[offset * indexMultiple_];
+    return values[offset];
   }
 
   exec_in_t readNullFree(vector_size_t offset) const {
     return operator[](offset);
   }
 
   bool isSet(vector_size_t offset) const {
-    return !rawNulls_ || !bits::isBitNull(rawNulls_, offset * indexMultiple_);
+    return !vector->isNullAt(offset);
```

Comment:
Any particular reason not to store rawNulls_ anymore?

Reply:
this is a new reader, so I did not change any previous behavior.